graphbolt/src/cuda/common.h

```diff
@@ -9,9 +9,63 @@
 #include <c10/cuda/CUDAException.h>
 #include <cuda_runtime.h>
 
+#include <memory>
+#include <unordered_map>
+
 namespace graphbolt {
 namespace cuda {
 
+/*
+How to use this class to get a nonblocking thrust execution policy that uses
+torch's memory pool and the current cuda stream:
+
+cuda::CUDAWorkspaceAllocator allocator(ctx);
+const auto stream = torch::cuda::getDefaultCUDAStream();
+const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
+
+Now, one can pass exec_policy to thrust functions
+
+To get an integer array of size 1000 whose lifetime is managed by unique_ptr,
+use: auto int_array = allocator.alloc_unique<int>(1000); int_array.get() gives
+the raw pointer.
+*/
+class CUDAWorkspaceAllocator {
```

Comment:
You are right, let me quickly fix that. Instead of returning a torch tensor, I can return a custom wrapper datatype that contains the type of storage we request.

Reply:
Actually, in order for this class to be an allocator and return raw pointer in its allocate function, it needs to hold onto the tensors it uses to allocate that storage. I can't come up with another solution.

Reply:
We can open an issue and improve it when a better solution is found, using a similar technique as done in tensoradapter whose link is given above.

Reply:
Example usage of this class in this file: https://github.com/dmlc/dgl/blob/master/src/array/cuda/labor_sampling.cu#L525

Reply:
As soon as this PR is merged, I will open an issue to track simplifying this class.